src/extractor/guidance/mergable_road_detector.cpp

```diff
@@ -0,0 +1,427 @@
+{
+    return [name_id, &node_based_graph](const MergableRoadDetector::MergableRoadData &road) {
+        // since we filter here, we don't want any other name than the one we are looking for
+        return name_id != node_based_graph.GetEdgeData(road.eid).name_id;
+    };
+}
+}
+
+MergableRoadDetector::MergableRoadDetector(const util::NodeBasedDynamicGraph &node_based_graph,
+                                           const std::vector<QueryNode> &node_coordinates,
+                                           const IntersectionGenerator &intersection_generator,
+                                           const CoordinateExtractor &coordinate_extractor,
+                                           const util::NameTable &name_table,
+                                           const SuffixTable &street_name_suffix_table)
+    : node_based_graph(node_based_graph), node_coordinates(node_coordinates),
+      intersection_generator(intersection_generator), coordinate_extractor(coordinate_extractor),
+      name_table(name_table), street_name_suffix_table(street_name_suffix_table)
+{
+}
+
+bool MergableRoadDetector::CanMergeRoad(const NodeID intersection_node,
+                                        const IntersectionShapeData &lhs,
+                                        const IntersectionShapeData &rhs) const
+{
+    // roads should be somewhat close
+    if (angularDeviation(lhs.bearing, rhs.bearing) > MERGABLE_ANGLE_DIFFERENCE)
+        return false;
+
+    const auto &lhs_edge_data = node_based_graph.GetEdgeData(lhs.eid);
+    const auto &rhs_edge_data = node_based_graph.GetEdgeData(rhs.eid);
+
+    // and they need to describe the same road
+    if (!RoadDataIsCompatible(lhs_edge_data, rhs_edge_data))
+        return false;
+
+    /* don't use any circular links, since they mess up detection we jump out early.
+     *
+     *          / -- \
+     * a ---- b - - /
+     */
+    const auto road_target = [this](const MergableRoadData &road) {
+        return node_based_graph.GetTarget(road.eid);
+    };
+
+    // TODO might have to skip over trivial intersections
+    if (road_target(lhs) == intersection_node || road_target(lhs) == intersection_node)
+        return false;
+
+    // Don't merge turning circles/traffic loops
+    if (IsTrafficLoop(intersection_node, lhs) || IsTrafficLoop(intersection_node, rhs))
+        return false;
+
+    // needs to be checked prior to link roads, since connections can seem like links
+    if (ConnectAgain(intersection_node, lhs, rhs))
+        return true;
+
+    // Don't merge link roads
+    if (IsLinkRoad(intersection_node, lhs) || IsLinkRoad(intersection_node, rhs))
+        return false;
+
+    // check if we simply split up prior to an intersection
+    if (IsNarrowTriangle(intersection_node, lhs, rhs))
+        return true;
+
+    // finally check if two roads describe the direction
+    return HaveSameDirection(intersection_node, lhs, rhs);
```

Comment:
At least I tried... It's not yet completely clear. There is a dedicated ticket for this, though: https://github.com/Project-OSRM/osrm-backend/issues/3473